Declarative Configuration - Updates to technical details after breaking changes (#86)
medic/medic-docs: configuration/developing-community-health-applications.md

| `context.expression` | A JavaScript expression which is evaluated when a contact profile or the reports tab is viewed. If the expression evaluates to true, the form will be listed as an available action. The inputs `contact`, `user`, and `summary` are available.

By default, forms are not shown on the reports tab, use `"expression": "!contact"` to show the form on the Reports tab since there is no contact for this scenario. | no |

In this example properties file, the associated form would only show on a person's page, and only if their sex is unspecified or female and they are between 10 and 65 years old:

```json
  {
    "title": [
      {
        "locale": "en",
        "content": "New Pregnancy"
      },
      {
        "locale": "hi",
        "content": "नई गर्भावस्था"
      }
    ],
    "icon": "pregnancy-1",
    "context": {
      "person": true,
      "place": false,
      "expression": "contact.type === 'person' && (!contact.sex || contact.sex === 'female') && (!contact.date_of_birth || (ageInYears(contact) >= 10 && ageInYears(contact) < 65))"
    }
  }
```

#### Build
    

> Forms are converted with the `convert-app-forms` or `convert-contact-forms` actions in `medic-conf`, and added to your application with the corresponding upload actions.

Reviewer: Just cleaning up while we are here for consistency.

```suggestion
Convert and build the app forms into your application using the `convert-app-forms` and `upload-app-forms` actions in `medic-conf`.
```

>     medic-conf --local convert-app-forms upload-app-form

Reviewer: ```suggestion
    medic-conf --local convert-app-forms upload-app-forms
```

| `resolvedIf` | `function(contact, report, event, dueDate)` | Return true to mark the task as "resolved". A resolved task uses memory on the phone, but is not displayed. | yes |
| `events` | Array of events | An event is used to specify the timing of the task. | yes |
| `events[n].id` | string | Can help as a descriptive name (eg `pregnancy-high-risk`). One task will appear per unique id, so re-using ids can be useful to avoid duplicate tasks appearing. | no |
| `events[n].days` | int | Number of days after the doc's `reported_date` that the event is due | yes, if `dueDate` is not set |
| `events[n].dueDate` | `function(event, contact, report)` | Returns a `Date` object for the day when this event is due. | yes, if `days` is not set |
| `events[n].start` | int | Number of days to show the task before it is due | yes |
| `events[n].end` | int | Number of days to show the task after it is due | yes |
| `actions` | Array of actions | The actions (forms) that a user can access after clicking on a task. If you put multiple forms here, the user will see a task summary screen where they can select which action they would like to complete. | yes |

> | `actions[n].type` | `'report'` or `'contact'` | When `'report'`, the action opens the given form. When `'contact'`, the action redirects to a contact's profile page. Defaults to 'report'. | no |

Author: I looked up the code and I'm pretty sure this is what is happening. I didn't tested it though. We could remove it if nobody ever uses it?

    medic-conf --local compile-app-settings backup-app-settings upload-app-settings 

> ## Contact Summaries

Reviewer: Even though this code is stored in `app_settings.contact_summary`, and the file for it is `contact-summary.templated.js`, I don't think _Contact Summaries_ defines this section well because it affects the context for the actions as well. What about _Contact Pages_, or _Contact Profiles_?

We need to remember to update the index as well if we change this from _Contacts_.

Author: I'm fine just reverting back to Contacts.

> In the `contact-summary.templated.js` file, write a script to control how a contact's profile page will appear by defining three outputs: `context`, `fields`, and `cards`.

Reviewer: We should add an intro about Contact Pages, like we do for Tasks and Targets. Also, I think we should shift away from "write a script" since we are describing the declarative format in the sections below, even if we don't say it directly.

```suggestion
Health workers can easily access and manage the families and people they serve as profile pages in the People tab. These contact profiles display basic information about the contact along with their history and upcoming tasks. You can control how a contact's profile page will appear and behave by defining these three outputs in the `contact-summary.templated.js` file: `context`, `fields`, and `cards`.
```

| `translate` | boolean | Whether or not to translate the value. Defaults to false. | no |
| `appliesIf` | boolean | True if the field should be shown. | no |
| `appliesToType` | Array of contact types | Filters the contacts for which `appliesIf` will be evaluated. For example, `['person']` or `['clinic', 'health_center']`. | no |
| `filter` | The display filter to apply to the value, eg: `{ value: '2005-10-09', filter: 'age' }` will render as "11 years". Common filters are: `age`, `phone`, `weeksPregnant`, `relativeDate`, `relativeDay`, `fullDate`, `simpleDate`, `simpleDateTime`, `lineage`, `resourceIcon`. For the complete list of filters, and more details on what each does, check out the code in [`medic/webapp/src/js/filters` dir](https://github.com/medic/medic/tree/master/webapp/src/js/filters). | no |

### Cards
An array of cards to show below the summary on the profile page. Each card has its own header and array of fields.

![Pregnancy card](img/pregnancy-card.png)
<!-- TODO: Add updated annotated screenshot of a profile card -->

| property | type | description | required |
|---|---|---|--|
| `label` | `translation key` | Label on top of card | yes |

> | `fields` | Array of `Fields` | The content of the card | yes |

Reviewer: I like the idea of consolidating to a common Field type. 

Are we sure that all fields in `Fields` are valid here? Eg `appliesToType`

Also, some available in this schema are not in `Fields` eg `fields.context`.

If we can reconcile those difference this makes a lot of sense.

Author: You're right. Some weird stuff happening...

Author: Okay - reverted and corrected.

| property | type | description | required |
|---|---|---|---|
| `label` | `translation key` | Label shown with the field | yes |
| `icon` | `string` | The name of the icon to display beside this field, as defined through the Configuration > Icons page. | no |
| `value` | `string` | The value shown for the field | yes |
| `filter` | `string` | The display filter to apply to the value, eg: `{ value: '2005-10-09', filter: 'age' }` will render as "11 years". Common filters are: `age`, `phone`, `weeksPregnant`, `relativeDate`, `relativeDay`, `fullDate`, `simpleDate`, `simpleDateTime`, `lineage`, `resourceIcon`. For the complete list of filters, and more details on what each does, check out the code in [`medic/webapp/src/js/filters` dir](https://github.com/medic/medic/tree/master/webapp/src/js/filters). | no |
| `width` | `integer` | The horizontal space for the field. Common values are 12 for full width, 6 for half width, or 3 for quarter width. Default 12 | no |
| `translate` | `boolean` | Whether or not to translate the value. Defaults to false. | no |

> | `context` | `object` | [Translation variables](https://angular-translate.github.io/docs/#/guide/06_variable-replacement) | no |

Reviewer: The word _context_ is used in many different ways across this documentation, and it is not clear what this `context` field does. We should add some ~context~ info to this `context` field, to make it clear whether or not someone would want to use this without going to the external documentation.

Reviewer: An example would be helpful for this too.

Author: Ya. I found that really confusing. I previously opened this https://github.com/medic/medic-conf/issues/169

> | `appliesIf` | `boolean` | True if the field should be shown. | no |

Reviewer: This one hasn't changed as part of this PR, but is missing the function prototype. Worth adding while we are here.

| property | type | description | required |
|---|---|---|--|
| `label` | `translation key` | Label on top of card | yes |
| `appliesToType` | Array of contact types `string` | Only calls `appliesIf` if the contact's type matches one or more of the array elements. For example, `['person']`. | no |
| `appliesIf` | `boolean` | True if the field should be shown. | no |
| `modifyContext` | `function(context)` | Used to modify or add values of the data which is passed as input to forms filled from the contact page | no |
| `fields` | `Array[]` of [fields](#fields) | The content of the card | yes |

> | `fields[n].appliesIf` | `boolean` or `function(report)` | Same as Fields.appliesIf above | |

Reviewer: When reading the markdown as a reference it is hard to follow by jumping back and forth with the Fields section. I would duplicate the info, and we can still note that this is the same as Fields.appliesIf above. Listing the required fields here is also useful without having to jump around in the doc.
